File: examples/csharp/SerialLink.cs
using System.IO.Ports;
using System.Threading.Channels;
using Mavlink;

namespace MavlinkSitlGcs;

/// <summary><see cref="MavlinkLink"/> implementation over a serial/COM port.</summary>
public sealed class SerialMavlinkLink : MavlinkLink
{
    private readonly SerialPort _port;
    private readonly Channel<byte[]> _receive = Channel.CreateUnbounded<byte[]>();
    private readonly CancellationTokenSource _cts = new();
    private readonly Task _readTask;
    private bool _closed;

    private SerialMavlinkLink(SerialPort port)
    {
        _port = port;
        _readTask = Task.Run(ReadLoopAsync);
    }

    /// <summary>Open <paramref name="portName"/> at <paramref name="baudRate"/> (SITL commonly uses 57600 or 115200).</summary>
    public static SerialMavlinkLink Open(string portName, int baudRate = 57600)
    {
        var port = new SerialPort(portName, baudRate, Parity.None, 8, StopBits.One)
        {
            DtrEnable = true,
            RtsEnable = true,
            ReadTimeout = 50,
        };
        port.Open();
        return new SerialMavlinkLink(port);
    }

    public override IAsyncEnumerable<byte[]> Receive => _receive.Reader.ReadAllAsync();

    public override Task SendAsync(byte[] data, CancellationToken cancellationToken = default)
    {
        if (_closed)
        {
            throw new InvalidOperationException("SerialMavlinkLink is closed");
        }

        _port.Write(data, 0, data.Length);
        return Task.CompletedTask;
    }

    public override async Task CloseAsync()
    {
        if (_closed)
        {
            return;
        }

        _closed = true;
        _cts.Cancel();

        if (_port.IsOpen)
        {
            _port.Close();
        }

        _receive.Writer.TryComplete();
        try
        {
            await _readTask.ConfigureAwait(false);
        }
        catch (OperationCanceledException)
        {
        }

        _cts.Dispose();
        _port.Dispose();
    }

    private async Task ReadLoopAsync()
    {
        var buffer = new byte[4096];

        while (!_closed && _port.IsOpen)
        {
            try
            {
                var bytesAvailable = _port.BytesToRead;
                if (bytesAvailable > 0)
                {
                    var count = _port.Read(buffer, 0, Math.Min(bytesAvailable, buffer.Length));
                    if (count > 0)
                    {
                        var chunk = new byte[count];
                        Array.Copy(buffer, chunk, count);
                        await _receive.Writer.WriteAsync(chunk, _cts.Token).ConfigureAwait(false);
                    }
                }
                else
                {
                    await Task.Delay(10, _cts.Token).ConfigureAwait(false);
                }
            }
            catch (OperationCanceledException) when (_closed || _cts.IsCancellationRequested)
            {
                break;
            }
            catch (Exception ex) when (!_closed)
            {
                _receive.Writer.TryComplete(ex);
                break;
            }
        }
    }
}

File: examples/typescript/src/serial_link.ts
import { SerialPort } from 'serialport';
import type { MavlinkLink } from '../../../generated/ts/protocols/mavlink_link.ts';

type ReceiveWaiter = {
  resolve: (value: IteratorResult<Uint8Array>) => void;
  reject: (error: unknown) => void;
};

/** [MavlinkLink] implementation over a cross-platform serial port. */
export class SerialMavlinkLink implements MavlinkLink {
  private readonly port: SerialPort;
  private readonly queue: Uint8Array[] = [];
  private readonly waiters: ReceiveWaiter[] = [];
  private closed = false;

  private constructor(port: SerialPort) {
    this.port = port;

    port.on('data', (data: Buffer) => {
      const chunk = new Uint8Array(data);
      const waiter = this.waiters.shift();
      if (waiter !== undefined) {
        waiter.resolve({ done: false, value: chunk });
        return;
      }
      this.queue.push(chunk);
    });

    port.on('error', (error) => {
      while (this.waiters.length > 0) {
        this.waiters.shift()!.reject(error);
      }
    });
  }

  /** Open [path] at [baudRate] (MAVLink SITL commonly uses 57600 or 115200). */
  static async open(
    path: string,
    baudRate = 57600,
  ): Promise<SerialMavlinkLink> {
    const port = new SerialPort({
      path,
      baudRate,
      dataBits: 8,
      parity: 'none',
      stopBits: 1,
      autoOpen: false,
    });

    await new Promise<void>((resolve, reject) => {
      port.open((error) => {
        if (error !== null && error !== undefined) {
          reject(error);
          return;
        }
        resolve();
      });
    });

    await new Promise<void>((resolve, reject) => {
      port.set({ dtr: true, rts: true }, (error) => {
        if (error !== null && error !== undefined) {
          reject(error);
          return;
        }
        resolve();
      });
    });

    return new SerialMavlinkLink(port);
  }

  readonly receive: AsyncIterable<Uint8Array> = {
    [Symbol.asyncIterator]: () => ({
      next: (): Promise<IteratorResult<Uint8Array>> => {
        if (this.closed && this.queue.length === 0) {
          return Promise.resolve({ done: true, value: undefined });
        }

        const queued = this.queue.shift();
        if (queued !== undefined) {
          return Promise.resolve({ done: false, value: queued });
        }

        return new Promise<IteratorResult<Uint8Array>>((resolve, reject) => {
          this.waiters.push({ resolve, reject });
        });
      },
    }),
  };

  async send(data: Uint8Array): Promise<void> {
    if (this.closed) {
      throw new Error('SerialMavlinkLink is closed');
    }

    await new Promise<void>((resolve, reject) => {
      this.port.write(Buffer.from(data), (error) => {
        if (error !== null && error !== undefined) {
          reject(error);
          return;
        }
        resolve();
      });
    });
  }

  async close(): Promise<void> {
    if (this.closed) {
      return;
    }
    this.closed = true;

    await new Promise<void>((resolve) => {
      this.port.close(() => resolve());
    });

    while (this.waiters.length > 0) {
      this.waiters.shift()!.resolve({ done: true, value: undefined });
    }
  }
}
